apps/examples/aifw_test/SineWaveProcessHandler.cpp

```diff
@@ -0,0 +1,114 @@
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied. See the License for the specific
+ * language governing permissions and limitations under the License.
+ *
+ ****************************************************************************/
+
+#include <memory>
+#include "aifw/aifw.h"
+#include "aifw/aifw_log.h"
+#include "aifw/aifw_utils.h"
+#include "aifw/AIDataBuffer.h"
+#include "SineWaveProcessHandler.h"
+
+SineWaveProcessHandler::SineWaveProcessHandler() : mRawData(NULL), mInvokeOutput(NULL)
+{
+}
+
+SineWaveProcessHandler::~SineWaveProcessHandler()
+{
+	if (mRawData) {
+		delete[] mRawData;
+		mRawData = NULL;
+	}
+	if (mInvokeOutput) {
+		delete[] mInvokeOutput;
+		mInvokeOutput = NULL;
+	}
+}
+
+AIFW_RESULT SineWaveProcessHandler::parseData(void *data, uint16_t count, float *parsedData, AIModelAttribute *modelAttribute)
+{
+	if (!data) {
+		AIFW_LOGE("raw data from source is NULL");
+		return AIFW_INVALID_ARG;
+	}
+	if (!parsedData) {
+		AIFW_LOGE("parsed data buffer argument is NULL");
+		return AIFW_INVALID_ARG;
+	}
+	/* Here raw data values are parsed. Required raw data values can be selected or new data can be created using different raw data values. */
+    for (uint16_t i = 0; i < modelAttribute->rawDataCount; i++) {
+    	parsedData[i] = ((float *)data)[i];
+    }
+	return AIFW_OK;
+}
+
+AIFW_RESULT SineWaveProcessHandler::preProcessData(std::shared_ptr<aifw::AIDataBuffer> buffer, float *invokeInput, AIModelAttribute *modelAttribute)
+{
+	if (!buffer) {
+		AIFW_LOGE("Invalid argument - input data buffer");
+		return AIFW_INVALID_ARG;
+	}
+	if (!invokeInput) {
+		AIFW_LOGE("Invalid argument - output data buffer");
+		return AIFW_INVALID_ARG;
+	}
+	if (!mRawData) {
+		mRawData = new float[modelAttribute->rawDataCount];
+		if (!mRawData) {
+			AIFW_LOGE("Memory Allocation failed - data read buffer");
+			return AIFW_NO_MEM;
+		}
+	}
+	/* Read latest row from AI data buffer. At this time the latest row includes only parsed raw data and does not include invoke result. */
+	AIFW_RESULT res = buffer->readData(mRawData, 0, modelAttribute->rawDataCount, 0);
+	if (res != AIFW_OK) {
+		AIFW_LOGE("Reading Data from the buffer failed. ret: %d", res);
+		return res;
+	}
+	/* Pre processing can be done at this point e.g. data can be normalized using mean and standard deviation values specific to model. Fnally pre-processed values, ready for inference is to be copied in invokeInput buffer */
+	memcpy(invokeInput, mRawData, modelAttribute->invokeInputCount * sizeof(float));
+	return AIFW_OK;
+}
+
+AIFW_RESULT SineWaveProcessHandler::postProcessData(std::shared_ptr<aifw::AIDataBuffer> buffer, float *resultData, AIModelAttribute *modelAttribute)
+{
+	if (!buffer) {
```

Comment:
AIDataBuffer should be valid for AI FW to work. This is application source code. AI FW can have checks before calling application code & this file is for illustration. Usually callee adds argument checks to avoid any issues later.
I understand these can be redundant but shouldn't we leave it for application developer to decide?

Reply:
My approach is that reduce unnecessary code as possible as we can.